src/v/storage/segment_deduplication_utils.cc

```diff
@@ -114,6 +119,23 @@ ss::future<model::offset> build_offset_map(
               seg->filename());
             break;
         }
+        segment_full_path idx_path = seg->path().to_compacted_index();
+        while (true) {
+            auto state = co_await internal::detect_compaction_index_state(
+              idx_path, cfg);
+            if (!internal::compacted_index_needs_rebuild(state)) {
+                break;
+            }
+            // Rebuilding the compaction index will take the read lock again,
+            // so release here.
+            read_lock.return_all();
+            co_await internal::rebuild_compaction_index(
+              *iter, stm_manager, cfg, probe, resources);
+
+            // Take the lock again before checking the compaction index state
+            // to avoid races with truncations while building the offset map.
+            read_lock = co_await seg->read_lock();
```

Comment:
is it possible that between the above code and acquiring the lock here something else could have modified the segment state in such a way that the code that follows would be invalid? i.e. can the compaction index be removed again?

Reply:
Yea it's possible, which is why we continue with the loop and check the state again while the lock is held